accepted/0000-subscription-matching-in-public-clouds.md

```diff
@@ -0,0 +1,121 @@
+
+
+## The output from a `virtual-host-gatherer` plugin:
+The required output from a plugin is a JSON response to the STDOUT, like the following
+
+```
+{
+    "my-amazon-ec2-region-X": {
+        "tendancy1": {
+            "hostIdentifier": "tenancy1",
+            ...
+            "type": "aws",
+            "vms": {
+                "my-aws-instance-1": "564d6d90-459c-2256-8f39-3cb2bd24b7b0"
+            }
+        },
+        "tenancy2": {
+            "hostIdentifier": "tenancy2",
+            ...
+            "type": "aws",
+            "vms": {
+                "my-aws-instance-2": "4230c60f-3f98-2a65-f7c3-600b26b79c22",
+                "my-aws-instance-3": "4230b00f-0b21-0e9d-dfde-6c7b06909d5f",
+                "my-aws-instance-4": "4230e924-b714-198b-348b-25de01482fd9"
+            }
+        }
+    }
+}
+```
+TODO: Complete the example JSON to add all attributes required from the Uyuni database.
+
+NOTE: All "vms" that belongs to the same virtual-host-manager, in the above example "my-amazon-ec2-region-X", regarding its tenancy, they all will be considered in the same virtualization group.
+
+## Add new Virtual Instance Types to the DB
+New virtual instance types would be needed on the `rhnVirtualInstanceType` database table, these would be `azure`, `aws`, `gce` and `generic` to indicate the type of the instance.
+
+
+## Gather instances and "uuid" from Public Clouds using API
+
+### Azure module:
+
+TODO: Examples of how to deal with Azure API
+
+The `vmId` reported from the API response corresponds with the "uuid" of the instance.
+
+### AWS module:
+
+TODO: Examples of how to deal with AWS API
+
+The `instanceId` returned from the API is **NOT** the "uuid", on this particular module, we could get the corresponding "uuid" from an "uuid" instance tag that needs to be set before in order to this module to incorporate this instance into the generated output.
+
+### Google Compute Engine module:
+
+TODO: Examples of how to deal with GCE API
+
+In the case of GCE, the instance id is **NOT** the "uuid" so we need to also to gather the "uuid" from a previously store "uuid" on the instanceMetadata.
+
+### A generic JSON-file module:
+
+TODO: Describe generic module
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+## Gathering the UUID for the virtual instances
+
+For AWS and GCE, the instance id that is returned from the API is not actually the real "uuid" from the instance. In order to get it, it would be previously needed to store that "uuid" on the metadata/tags for each instance.
+
+TODO: Evaluate if the "uuid" could be stored from within the instance into the "instanceMetadata" / "instanceTags" using the internal API available for the instances. That way, an automate action could be executed on the registered instances to store this information and then have it available when running `virtual-host-gatherer`.
```

Comment:
AFAIK this means we need to try to change the registration process (salt and traditional) to put a different "string" into the uuid field. We just need to be able to get the identifier from inside of the vm.
I think there was no need in SUSE Manager that it must really look like an uuid. We may put any string into it which uniquely identify the instance.

Reply:
You're right, and I think using directly the provider "instance id" instead a real smbios "uuid" is the smoother way here. A minor change is needed on the Java side to allow providing "uuid" which are not real "uuid" but this is also covered on the RFC.

I've pushed some new changes to reflect this.